celerpy/visualize.py

```diff
@@ -0,0 +1,352 @@
+
+    def __init__(self, ids):
+        """Give a list of unique IDs that will be remapped to their position in
+        this list.
+        """
+        assert ids.size > 0
+        fltids = np.array(ids, dtype=float)
+        bounds = np.concatenate([fltids - 0.5, [fltids[-1] + 0.5]])
+        super().__init__(bounds, len(bounds) - 1, clip=False)
+        self.vmin = 0
+        self.vmax = ids[-1]
+
+        # Remap matids to ID index: array up to the last valid matid.
+        ids_to_idx = np.zeros((ids[-1] + 1), dtype=np.int16)
+        ids_to_idx[ids] = np.arange(len(ids), dtype=np.int16)
+        self.ids_to_idx = ids_to_idx
+
+    def __call__(self, value, clip=None):
+        """Convert IDs to indices.
+
+        For integer values, this is a much faster operation than the base
+        class's implementation of __call__.
+        """
+        if not np.issubdtype(value.dtype, np.integer):
+            # When rendering color bars, matplotlib passes an array of
+            # floating-point midpoint values. Let the base class logic (slower
+            # but compatible with the colorbar mechanics) handle it.
+            return super().__call__(value, clip)
+
+        mask = np.ma.getmaskarray(value)
+        invalid = value >= self.ids_to_idx.size
+        value[invalid] = 0
+        mask |= invalid
+
+        result = self.ids_to_idx[value]
+        result = np.ma.array(result, mask=mask)
+        return result
+
+
+def calc_image_axes(image: model.ImageParams):
+    """Calculate label/min/max for x and y axes from an image result."""
+    down = np.array(image.down)
+    right = np.array(image.right)
+    height = image.pixel_width * image.dims[0]
+    width = image.pixel_width * image.dims[1]
+    upper_left = image.origin
+    lower_left = down * height + upper_left
+    units = UNIT_LENGTH[image.units]
+
+    def calc_axes(length, dir):
+        for d, dimname in enumerate("xyz"):
+            if 1 - abs(dir[d]) < 1e-6:
+                lower = lower_left[d]
+                upper = lower + length * dir[d]
+                label = f"{dimname} [{units}]"
+                break
+        else:
+            # No orthogonal axis found
+            label = "Position along ({}) from {} [{}]".format(
+                ",".join("{:.3f}".format(v if v != 0 else 0) for v in dir),
+                lower_left,
+                units,
+            )
+            (lower, upper) = (0, length)
+        return LabeledAxis(label, lower, upper)
+
+    return LabeledAxes(x=calc_axes(width, right), y=calc_axes(height, -down))
+
+
+class Imager:
+    axes: Optional[LabeledAxes] = None
+
+    def __init__(self, celer_geo, image: model.ImageInput):
+        self.celer_geo = celer_geo
+        self.image = image
+        self.axes = None  # Lazily update
+
+    def __call__(
+        self,
+        ax,
+        geometry: Optional[model.GeometryEngine] = None,
+        memspace: Optional[model.MemSpace] = None,
+        colorbar=None,
+    ):
```

Comment:
return type annotation

Reply:
What does one do when the return type is an implementation detail (it's a result from matplotlib)? In C++ I'd use 'auto'...

Reply:
For this case it looks like it will return a `matplotlib.image.AxesImage`

Reply:
Actually it looks like I mistakenly was just returning `im` when I meant to return a `dict` 😬 so good catch with the return type